source/pele/cell_lists.h

```diff
@@ -150,40 +200,273 @@ namespace pele {
+            }
+
+            ll1 = lower_left1;
+            ll2 = lower_left2;
+            ll1[i] += rcell;
+            ll2[i] += rcell;
+            m_dist->get_rij(dr.data(), ll1.data(), ll2.data());
+            dri = std::abs(dr[i]);
+            if (dri < min_dist) {
+                min_dist = dri;
+            }
+
+            minimum_dist[i] = min_dist;
+        }
+        double r2_min = dot<ndim> (minimum_dist, minimum_dist);
+        return std::sqrt(r2_min);
+    }
+
+    /**
+     * recursive function to find the neighbors of a given cell
+     */
+    size_t find_neighbors(size_t idim, cell_vec_t v0,
+            std::vector<size_t> & neighbors,
+            cell_vec_t const & vorigin
+            ) const
+    {
+        if (idim == ndim) {
+            double rmin = minimum_distance(v0, vorigin);
+            if (rmin <= m_rcut) {
+                neighbors.push_back(to_index(v0));
+//                std::cout << "  adding neighbor    "<< v0 << " rmin " << rmin << "\n";
+                return 1;
+            }
+//            std::cout << "  rejecting neighbor "<< v0 << " rmin " << rmin << "\n";
+            return 0;
+        }
+//        std::cout << "find neighbors " << idim << " " << v0 << "\n";
+        size_t nfound = 0;
+        size_t n;
+        auto v = v0;
+        long max_negative = (m_ncells_vec[idim] - 1) / 2;
+        long max_positive = m_ncells_vec[idim] / 2;
+        // step in the negative direction
+        long offset = 0;
+        while (offset <= max_negative) {
+            v[idim] = v0[idim] - offset;
+            n = find_neighbors(idim+1, v, neighbors, vorigin);
+            if (n == 0) break;
+            nfound += n;
+            ++offset;
+        }
+        // step in the positive direction
+        offset = 1;
+        while (offset <= max_positive) {
+            v[idim] = v0[idim] + offset;
+            n = find_neighbors(idim+1, v, neighbors, vorigin);
+            if (n == 0) break;
+            nfound += n;
+            ++offset;
+        }
+
+        return nfound;
+    }
+
+    /**
+     * return a vector of all the neighbors of icell (including icell itself)
+     */
+    std::vector<size_t> find_all_neighbors(size_t icell) const
+    {
+        auto vcell = to_cell_vec(icell);
 
+        std::vector<size_t> neighbors;
+        find_neighbors(0, vcell, neighbors, vcell);
+//        std::cout << pele::Array<size_t>(neighbors) << std::endl;
+        return neighbors;
+    }
+
+    /**
+     * return a list of all pairs of neighboring cells
+     *
+     * The algorithm could be improved.  If the distance function is periodic
+     * then each cell has the same structure of neighbors.  We could just keep
+     * a list of cell vector offsets and apply these to each cell to find the
+     * list of neighbor pairs.
+     */
+    void find_neighbor_pairs(std::vector<std::pair<size_t, size_t> > & cell_neighbors) const
+    {
+        cell_neighbors.reserve(m_ncells * std::pow(3, ndim));
```

Comment:
Because std::pow was somehow surprisingly slow with integer exponents, we put in the explicit loop instead in the old version.
But I guess speed is not that important here.